File: module_3/java/GenerateParenthesis.java
/*
 *   Title: Generate parentheses
 *
 *   Problem:
 *      Given n pairs of parentheses, write a function to generate all
 *      combinations of well-formed parentheses.  For example, given n = 3, a
 *      solution set is:
 *      [
 *        "((()))",
 *         "(()())",
 *         "(())()",
 *         "()(())",
 *         "()()()"
 *      ]
 *
 *   Execution: javac GenerateParenthesis.java && java GenerateParenthesis
 */
import java.util.*;


public class GenerateParenthesis {
    public static List<String> generateParenthesisBruteForce(int n) {
        List<String> combinations = new ArrayList();
        generateAll(new char[2 * n], 0, combinations);
        return combinations;
    }

    public static void generateAll(char[] current, int pos, List<String> result) {
        if (pos == current.length) {
            if (valid(current))
                result.add(new String(current));
        } else {
            current[pos] = '(';
            generateAll(current, pos+1, result);
            current[pos] = ')';
            generateAll(current, pos+1, result);
        }
    }

    public static boolean valid(char[] current) {
        int balance = 0;
        for (char c: current) {
            if (c == '(') balance++;
            else balance--;
            if (balance < 0) return false;
        }
        return (balance == 0);
    }

    public static List<String> generateParenthesisBacktracking(int n) {
        List<String> ans = new ArrayList();
        backtrack(ans, "", 0, 0, n);
        return ans;
    }

    public static void backtrack(List<String> ans, String cur, int open, int close, int max){
        if (cur.length() == max * 2) {
            ans.add(cur);
            return;
        }

        if (open < max)
            backtrack(ans, cur+"(", open+1, close, max);
        if (close < open)
            backtrack(ans, cur+")", open, close+1, max);
    }
    public static void main(String[] args) {
        
        List<String> list = new ArrayList<String>();
        list.add("((()))");
        list.add("(()())");
        list.add("(())()");
        list.add("()(())");
        list.add("()()()");

        // Brute-force approach:
        assert generateParenthesisBruteForce(3) == list;

        // Backtracking approach:
        assert generateParenthesisBacktracking(3) == list;

        System.out.println("Passed all test cases");
    }
    
}

File: module_3/python/generate_parenthesis.py
"""
Title: Generate parenthesis

Problem:
    Given n pairs of parentheses, write a function to generate all combinations of well-formed parentheses.
    For example, given n = 3, a solution set is:
        [
          "((()))",
          "(()())",
          "(())()",
          "()(())",
          "()()()"
        ]

Execution: python generate_parenthesis.py
"""
import unittest
from typing import List


def generate_parenthesis_brute_force(n: int) -> List[str]:
    """Brute-force generate parenthesis."""
    def generate(A=[]):
        if len(A) == 2*n:
            if valid(A):
                ans.append("".join(A))
        else:
            A.append('(')
            generate(A)
            A.pop()
            A.append(')')
            generate(A)
            A.pop()

    def valid(A):
        bal = 0
        for c in A:
            if c == '(': bal += 1
            else: bal -= 1
            if bal < 0: return False
        return bal == 0

    ans = []
    generate()

    return ans


def generate_parenthesis_backtracking(n: int) -> List[str]:
    """Backtracking generate parenthesis."""
    ans = []

    def backtrack(S: str = "", left: int = 0, right: int = 0):
        if len(S) == 2 * n:
            ans.append(S)
            return
        if left < n:
            backtrack(S+'(', left+1, right)
        if right < left:
            backtrack(S+')', left, right+1)

    backtrack()
    return ans


class TestGenerateParenthesis(unittest.TestCase):
    """Unit test for generate_parenthesis."""

    def test_1(self):
        """Test for input shown below."""
        test_output = [
            "((()))",
            "(()())",
            "(())()",
            "()(())",
            "()()()"]
        self.assertEqual(generate_parenthesis_brute_force(3), test_output)
        self.assertEqual(generate_parenthesis_backtracking(3), test_output)


if __name__ == '__main__':
    unittest.main()
